### PR TITLE
merge master into migration

### PR DESCRIPTION
**Pull request by jessfraz**
This merges the master branch into the migration branch and fixed conflicts.

### CODE REVIEW

File: parts/k8s/kubernetesmastervars.t [hunk 1]
```diff
@@ -131,15 +207,16 @@
...
...ipalClientId'),' SERVICE_PRINCIPAL_CLIENT_SECRET=',variables('servicePrincipalClientSecret'),' KUBELET_PRIVATE_KEY=',variables('clientPrivateKey'),' TARGET_ENVIRONMENT=',variables('targetEnvironment'),' NETWORK_POLICY=',variables('networkPolicy'),' FQDNSuffix=',variables('fqdnEndpointSuffix'),' VNET_CNI_PLUGINS_URL=',variables('vnetCniLinuxPluginsURL'),' CNI_PLUGINS_URL=',variables('cniPluginsURL'),' MAX_PODS=',variables('maxPods'),' CLOUDPROVIDER_BACKOFF=',variables('cloudProviderBackoff'),' CLOUDPROVIDER_BACKOFF_RETRIES=',variables('cloudProviderBackoffRetries'),' CLOUDPROVIDER_BACKOFF_EXPONENT=',variables('cloudProviderBackoffExponent'),' CLOUDPROVIDER_BACKOFF_DURATION=',variables('cloudProviderBackoffDuration'),' CLOUDPROVIDER_BACKOFF_JITTER=',variables('cloudProviderBackoffJitter'),' CLOUDPROVIDER_RATELIMIT=',variables('cloudProviderRatelimit'),' CLOUDPROVIDER_RATELIMIT_QPS=',variables('cloudProviderRatelimitQPS'),' CLOUDPROVIDER_RATELIMIT_BUCKET=',variables('cloudProviderRatelimitBucket'),' USE_MANAGED_IDENTITY_EXTENSION=',variables('useManagedIdentityExtension'),' USE_INSTANCE_METADATA=',variables('useInstanceMetadata'),' CONTAINER_RUNTIME=',variables('containerRuntime'))]",
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastervars.t]: this reverted the changes in yangl900.

Comment (jessfraz) [reply parts/k8s/kubernetesmastervars.t]: so it still has all the variables from the original migration branch, but it added all the ones from master too, was there a reason you all had the others deleted?

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 2]
```diff
@@ -132,10 +146,61 @@ New-InfraContainer()
...
+        Set-AzureNetworkPolicy
+    }
+}
+
 function
 Write-KubernetesStartFiles($podCIDR)
 {
-    $KubeletArgList = @("--hostname-override=`$global:AzureHostname","--pod-infra-container-image=kubletwin/pause","--resolv-conf=""""""""","--kubeconfig=c:\k\config")
+    mkdir $global:VolumePluginDir
+    $KubeletArgList = @("--hostname-override=`$global:AzureHostname","--pod-infra-container-image=kubletwin/pause","--resolv-conf=""""""""","--kubeconfig=c:\k\config","--cloud-provider=azure","--cloud-config=c:\k\azure.json")
```
Comment (wenwu449) [inline parts/k8s/kuberneteswindowssetup.ps1]: cloud-config shall not be add back

File: parts/k8s/kubernetesmastervars.t (hunk 1 above)
Comment (wenwu449) [reply parts/k8s/kubernetesmastervars.t]: those parameters grant too much permission to agent nodes, basically they can access ARM resources and do infrastructure level change.

**Comment by jessfraz** [comment]
Fixed so it just applies the clear containers patches :)

File: pkg/api/vlabs/validate_test.go [hunk 3]
```diff
@@ -506,3 +506,39 @@ func getK8sDefaultProperties() *Properties {
...
 }
+
+func Test_Properties_ValidateContainerRuntime(t *testing.T) {
+	p := &Properties{}
+	p.OrchestratorProfile = &OrchestratorProfile{}
+	p.OrchestratorProfile.OrchestratorType = Kubernetes
+
+	for _, policy := range ContainerRuntimeValues {
+		p.OrchestratorProfile.KubernetesConfig = &KubernetesConfig{}
+		p.OrchestratorProfile.KubernetesConfig.NetworkPolicy = policy
```
Comment (wenwu449) [inline pkg/api/vlabs/validate_test.go]: .ContainerRuntime

File: pkg/api/vlabs/validate_test.go [hunk 4]
```diff
@@ -506,3 +506,39 @@ func getK8sDefaultProperties() *Properties {
 		},
 	}
 }
+
+func Test_Properties_ValidateContainerRuntime(t *testing.T) {
+	p := &Properties{}
+	p.OrchestratorProfile = &OrchestratorProfile{}
+	p.OrchestratorProfile.OrchestratorType = Kubernetes
+
+	for _, policy := range ContainerRuntimeValues {
```
Comment (wenwu449) [inline pkg/api/vlabs/validate_test.go]: runtime

File: parts/k8s/kubernetesmastervars.t [hunk 5]
```diff
@@ -171,7 +172,7 @@
 {{end}}
     "provisionScript": "{{GetKubernetesB64Provision}}",
     "mountetcdScript": "{{GetKubernetesB64Mountetcd}}",
-    "provisionScriptParametersCommon": "[concat('KUBELET_PRIVATE_KEY=',variables('clientPrivateKey'),' NETWORK_POLICY=',variables('networkPolicy'),' APISERVER_PUBLIC_KEY=',variables('apiserverCertificate'),' MAX_PODS=',variables('maxPods'))]",
+    "provisionScriptParametersCommon": "[concat('KUBELET_PRIVATE_KEY=',variables('clientPrivateKey'),' NETWORK_POLICY=',variables('networkPolicy'),' APISERVER_PUBLIC_KEY=',variables('apiserverCertificate'),' MAX_PODS=',variables('maxPods'),' CONTAINER_RUNTIME=',variables('containerRuntime'))]",
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastervars.t:175]: does this mean master will also use clear-container runtime?

Comment (jessfraz) [reply parts/k8s/kubernetesmastervars.t:175]: The script checks if the node has nested virt support so I have jsut been using non nested virt nodes for the master

Comment (wenwu449) [reply parts/k8s/kubernetesmastervars.t:175]: awesome! Thanks for clarification

**Pull request merged**